test/kafka-upsert-benchmark/mzworkflows.py

```diff
@@ -0,0 +1,165 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import time
+from typing import List
+
+import pg8000
+
+from materialize import ui
+from materialize.mzcompose import (
+    Kafka,
+    Kgen,
+    Materialized,
+    Testdrive,
+    Workflow,
+    WorkflowArgumentParser,
+    Zookeeper,
+    say,
+)
+from materialize.ui import UIError
+
+mz_persist = Materialized(
+    name="materialized_persist",
+    hostname="materialized",
+    options="--persistent-user-tables --persistent-kafka-upsert-source --disable-persistent-system-tables-test",
+)
+
+mz_default = Materialized()
+
+prerequisites = ["zookeeper", "kafka"]
+
+services = [
+    Zookeeper(),
+    Kafka(),
+    mz_default,
+    mz_persist,
+    # This instance of Testdrive is used to initialize the benchmark, and needs to have a
+    # known seed.
+    Testdrive(no_reset=True, seed=1),
+    Kgen(),
+]
+
+
+def kafka_upsert_benchmark_step(
+    w: Workflow,
+    mz_service: str,
+    restart_mz: bool,
+    records_sent: int,
+    num_records: int,
+    num_keys: int,
+    value_bytes: int,
+    timeout_secs: int,
+):
+    if restart_mz:
+        w.kill_services(services=[mz_service], signal="SIGKILL")
+    w.run_service(
+        service="kgen",
+        command=" ".join(
+            [
+                f"--num-records={num_records}",
+                "--keys=random",
+                "--key-min=0",
+                f"--key-max={num_keys}",
+                "--values=bytes",
+                f"--max-message-size={value_bytes+1}",
+                f"--min-message-size={value_bytes}",
+                "--topic=testdrive-load-test-1",
+                "--quiet",
+            ]
+        ),
+    )
+
+    if restart_mz:
+        w.start_services(services=[mz_service])
+        w.wait_for_mz(service=mz_service)
+    expected = records_sent + num_records
+
+    host_port = int(w.composition.find_host_ports(mz_service)[0])
+    query = "SELECT * FROM num_records_ingested"
+    ui.progress(f"waiting for materialize to handle {query!r}", "C")
+    error = None
+    start_time = time.monotonic()
+
+    # Default to polling Materialize once a second because that is the current default
+    # timestamp frequency and polling any more frequently doesn't give new data.
+    for remaining in ui.timeout_loop(timeout_secs):
+        try:
+            conn = pg8000.connect(
+                host="localhost", port=host_port, user="materialize", timeout=1
+            )
+            conn.autocommit = True
+
+            cursor = conn.cursor()
+            cursor.execute("SELECT * FROM num_records_ingested")
```

Comment:
You have put the text of the SQL query in the `query` variable above, you may wish to reuse it here.